README.md
# SYNOPSIS 
[![NPM Package](https://img.shields.io/badge/npm-v1.1.3-blue.svg)](https://www.npmjs.org/package/moacjs-tx)
[![Build Status](https://travis-ci.org/wanpixiaozi/moacjs-tx.svg?branch=master)](https://travis-ci.org/wanpixiaozi/moacjs-tx)
[![Coverage Status](https://coveralls.io/repos/github/wanpixiaozi/moacjs-tx/badge.svg?branch=master)](https://coveralls.io/github/wanpixiaozi/moacjs-tx?branch=master)
[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)  

# INSTALL
`npm install moacjs-tx`

# USAGE

  - [example](https://github.com/wanpixiaozi/moacjs-tx/blob/master/examples/transaction.js)

```javascript
`const Chain3 = require('chain3')
const chain3 = new Chain3(new Chain3.providers.HttpProvider('http://localhost:8545')) 
const MoacTx = require('moacjs-tx')
const privateKey = XXX

const txParams = {
  nonce: '0x00',
  gasPrice: '0x09184e72a000', 
  gasLimit: '0x2710',
  to: '0x0000000000000000000000000000000000000000', 
  value: '0x00', 
  data: '0x7f7465737432000000000000000000000000000000000000000000000000000000600057',
  shardingFlag: 0,
  // chainId - mainnet: 99, testnet: 101
  chainId: 101
}

const tx = new MoacTx(rawTx)
const serializedTx = tx.sign(fromSecret)
chain3.mc.sendRawTransaction(serializedTx, function (err, hash) {
    if (!err) {
      console.log('succeed: ', hash)
      return hash
    } else {
      console.log('error:', err)
      console.log('raw tx:', rawTx)
    }
})`
````

**Note:** this package expects ECMAScript 6 (ES6) as a minimum environment. From browsers lacking ES6 support, please use a shim (like [es6-shim](https://github.com/paulmillr/es6-shim)) before including any of the builds from this repo.


# BROWSER  
For a browser build please get from https://github.com/wanpixiaozi/moacjs-tx/blob/master/dist/moacjs-tx.min.js.

# API
[./docs/](./docs/index.md)

# LICENSE
[MPL-2.0](https://tldrlegal.com/license/mozilla-public-license-2.0-(mpl-2))


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) 
(modification: no type change headlines) and this project adheres to 
[Semantic Version](http://semver.org/spec/v2.0.0.html).

## [1.1.3] - 2018-12-19
- update wording

## [1.1.2] - 2018-12-11
- fix errors

## [1.1.1] - 2018-12-10
- tune code, make it simple

## [1.1.0] - 2018-12-07
- Modify sign process refer to chain3 process, combine serialize process into sign function

## [1.0.0] - 2018-12-05
- First version for Moac blockchain


test/api.js
const tape = require('tape')
const utils = require('moacjs-util')
const rlp = utils.rlp
const Transaction = require('../index.js')
const txFixtures = require('./txs.json')
const txFixturesEip155 = require('./ttTransactionTestEip155VitaliksTests.json')
tape('[Transaction]: Basic functions', function (t) {
  var transactions = []

  t.test('should decode transactions', function (st) {
    txFixtures.slice(0, 3).forEach(function (tx) {
      var pt = new Transaction(tx.raw)
      st.equal('0x' + pt.nonce.toString('hex'), tx.raw[0])
      st.equal('0x' + pt.gasPrice.toString('hex'), tx.raw[1])
      st.equal('0x' + pt.gasLimit.toString('hex'), tx.raw[2])
      st.equal('0x' + pt.to.toString('hex'), tx.raw[3])
      st.equal('0x' + pt.value.toString('hex'), tx.raw[4])
      st.equal('0x' + pt.v.toString('hex'), tx.raw[6])
      st.equal('0x' + pt.r.toString('hex'), tx.raw[7])
      st.equal('0x' + pt.s.toString('hex'), tx.raw[8])
      st.equal('0x' + pt.data.toString('hex'), tx.raw[5])
      transactions.push(pt)
    })
    st.end()
  })

  t.skip('should serialize', function (st) {
    transactions.forEach(function (tx) {
      st.deepEqual(tx.serialize(), rlp.encode(tx.raw))
    })
    st.end()
  })

  t.test('should hash', function (st) {
    var tx = new Transaction(txFixtures[2].raw)
    st.deepEqual(tx.hash(), new Buffer('375a8983c9fc56d7cfd118254a80a8d7403d590a6c9e105532b67aca1efb97aa', 'hex'))
    st.deepEqual(tx.hash(false), new Buffer('61e1ec33764304dddb55348e7883d4437426f44ab3ef65e6da1e025734c03ff0', 'hex'))
    st.deepEqual(tx.hash(true), new Buffer('375a8983c9fc56d7cfd118254a80a8d7403d590a6c9e105532b67aca1efb97aa', 'hex'))
    st.end()
  })

  t.test('should hash with defined chainId', function (st) {
    var tx = new Transaction(txFixtures[3].raw)
    st.equal(tx.hash().toString('hex'), '0f09dc98ea85b7872f4409131a790b91e7540953992886fc268b7ba5c96820e4')
    st.equal(tx.hash(true).toString('hex'), '0f09dc98ea85b7872f4409131a790b91e7540953992886fc268b7ba5c96820e4')
    st.equal(tx.hash(false).toString('hex'), 'f97c73fdca079da7652dbc61a46cd5aeef804008e057be3e712c43eac389aaf0')
    st.end()
  })

  t.test('should verify Signatures', function (st) {
    transactions.forEach(function (tx) {
      st.equals(tx.verifySignature(), true)
    })
    st.end()
  })

  t.test('should not verify Signatures', function (st) {
    transactions.forEach(function (tx) {
      tx.s = utils.zeros(32)
      st.equals(tx.verifySignature(), false)
    })
    st.end()
  })

  t.test('should give a string about not verifying Signatures', function (st) {
    transactions.forEach(function (tx) {
      st.equals(tx.validate(true).slice(0, 54), 'Invalid Signature gas limit is too low. Need at least ')
    })
    st.end()
  })

  t.test('should validate', function (st) {
    transactions.forEach(function (tx) {
      st.equals(tx.validate(), false)
    })
    st.end()
  })

  t.test('should sign tx', function (st) {
    transactions.forEach(function (tx, i) {
      if (txFixtures[i].privateKey) {
        var privKey = new Buffer(txFixtures[i].privateKey, 'hex')
        tx.sign(privKey)
      }
    })
    st.end()
  })

  t.test("should get sender's address after signing it", function (st) {
    transactions.forEach(function (tx, i) {
      if (txFixtures[i].privateKey) {
        st.equals(tx.getSenderAddress().toString('hex'), txFixtures[i].sendersAddress)
      }
    })
    st.end()
  })

  t.test("should get sender's public key after signing it", function (st) {
    transactions.forEach(function (tx, i) {
      if (txFixtures[i].privateKey) {
        st.equals(tx.getSenderPublicKey().toString('hex'),
          utils.privateToPublic(new Buffer(txFixtures[i].privateKey, 'hex')).toString('hex'))
      }
    })
    st.end()
  })

  t.test("should get sender's address after signing it (second call should be cached)", function (st) {
    transactions.forEach(function (tx, i) {
      if (txFixtures[i].privateKey) {
        st.equals(tx.getSenderAddress().toString('hex'), txFixtures[i].sendersAddress)
        st.equals(tx.getSenderAddress().toString('hex'), txFixtures[i].sendersAddress)
      }
    })
    st.end()
  })

  t.skip('should verify signing it', function (st) {
    transactions.forEach(function (tx, i) {
      if (txFixtures[i].privateKey) {
        st.equals(tx.verifySignature(), true)
      }
    })
    st.end()
  })

  t.skip('should validate with string option', function (st) {
    transactions.forEach(function (tx) {
      tx.gasLimit = 30000
      st.equals(tx.validate(true), '')
    })
    st.end()
  })

  t.test('should round trip decode a tx', function (st) {
    var tx = new Transaction()
    tx.value = 5000
    var s1 = tx.serialize().toString('hex')
    var tx2 = new Transaction(s1)
    var s2 = tx2.serialize().toString('hex')
    st.equals(s1, s2)
    st.end()
  })

  t.test('should accept lesser r values', function (st) {
    var tx = new Transaction()
    tx.r = '0x0005'
    st.equals(tx.r.toString('hex'), '05')
    st.end()
  })

  t.test('should return data fee', function (st) {
    var tx = new Transaction()
    st.equals(tx.getDataFee().toNumber(), 0)

    tx = new Transaction(txFixtures[2].raw)
    st.equals(tx.getDataFee().toNumber(), 2496)

    st.end()
  })

  t.test('should return base fee', function (st) {
    var tx = new Transaction()
    st.equals(tx.getBaseFee().toNumber(), 53000)
    st.end()
  })

  t.test('should return upfront cost', function (st) {
    var tx = new Transaction({
      gasPrice: 1000,
      gasLimit: 10000000,
      value: 42
    })
    st.equals(tx.getUpfrontCost().toNumber(), 10000000042)
    st.end()
  })

  t.skip('Verify EIP155 Signature based on Vitalik\'s tests', function (st) {
    txFixturesEip155.forEach(function (tx) {
      var pt = new Transaction(tx.rlp)
      st.equal(pt.hash(false).toString('hex'), tx.hash)
      st.equal('0x' + pt.serialize().toString('hex'), tx.rlp)
      st.equal(pt.getSenderAddress().toString('hex'), tx.sender)
    })
    st.end()
  })

  t.skip('sign tx with chainId specified in params', function (st) {
    var tx = new Transaction({ chainId: 99 })
    st.equal(tx.getChainId(), 99)
    var privKey = new Buffer(txFixtures[0].privateKey, 'hex')
    tx.sign(privKey)
    var serialized = tx.serialize()
    var reTx = new Transaction(serialized)
    st.equal(reTx.verifySignature(), true)
    st.equal(reTx.getChainId(), 99)
    st.end()
  })

  t.test('allow chainId more than 1 byte', function (st) {
    var tx = new Transaction({ chainId: 0x16b2 })
    st.equal(tx.getChainId(), 0x16b2)
    st.end()
  })
})


test/fake.js
const tape = require('tape')
const utils = require('moacjs-util')
const FakeTransaction = require('../fake.js')

// Use private key 0x0000000000000000000000000000000000000000000000000000000000000001 as 'from' Account
var txData = {
  data: '0x7cf5dab00000000000000000000000000000000000000000000000000000000000000005',
  gasLimit: '0x15f90',
  gasPrice: '0x1',
  nonce: '0x01',
  to: '0xd9024df085d09398ec76fbed18cac0e1149f50dc',
  value: '0x0',
  from: '0x7e5f4552091a69125d5dfcb7b8c2659029395bdf',
  v: '0x1c',
  r: '0x25641558260ac737ea6d800906c6d085a801e5e0f0952bf93978d6fa468fbdfe',
  s: '0x5d0904b8f9cfc092805df0cde2574d25e2c5fc28907a9a4741b3e857b68b0778'
}

tape('[FakeTransaction]: Basic functions', function (t) {
  t.test('instantiate with from / create a hash', function (st) {
    st.plan(3)
    var tx = new FakeTransaction(txData)
    var hash = tx.hash()
    var cmpHash = Buffer.from('f74b039f6361c4351a99a7c6a10867369fe6701731d85dc07c15671ac1c1b648', 'hex')
    st.deepEqual(hash, cmpHash, 'should create hash with includeSignature=true (default)')
    var hash2 = tx.hash(false)
    var cmpHash2 = Buffer.from('0401bf740d698674be321d0064f92cd6ebba5d73d1e5e5189c0bebbda33a85fe', 'hex')
    st.deepEqual(hash2, cmpHash2, 'should create hash with includeSignature=false')
    st.notDeepEqual(hash, hash2, 'previous hashes should be different')
  })

  t.test('instantiate without from / create a hash', function (st) {
    var txDataNoFrom = Object.assign({}, txData)
    delete txDataNoFrom['from']
    st.plan(3)
    var tx = new FakeTransaction(txDataNoFrom)
    var hash = tx.hash()
    var cmpHash = Buffer.from('80a2ca70509414908881f718502e6bbb3bc67f416abdf972ea7c0888579be7b9', 'hex')
    st.deepEqual(hash, cmpHash, 'should create hash with includeSignature=true (default)')
    var hash2 = tx.hash(false)
    var cmpHash2 = Buffer.from('0401bf740d698674be321d0064f92cd6ebba5d73d1e5e5189c0bebbda33a85fe', 'hex')
    st.deepEqual(hash2, cmpHash2, 'should create hash with includeSignature=false')
    st.notDeepEqual(hash, hash2, 'previous hashes should be different')
  })

  t.test('should not produce hash collsions for different senders', function (st) {
    st.plan(1)
    var txDataModFrom = Object.assign({}, txData, { from: '0x2222222222222222222222222222222222222222' })
    var tx = new FakeTransaction(txData)
    var txModFrom = new FakeTransaction(txDataModFrom)
    var hash = utils.bufferToHex(tx.hash())
    var hashModFrom = utils.bufferToHex(txModFrom.hash())
    st.notEqual(hash, hashModFrom, 'FakeTransactions with different `from` addresses but otherwise identical data should have different hashes')
  })

  t.test('should retrieve "from" from signature if transaction is signed', function (st) {
    var txDataNoFrom = Object.assign({}, txData)
    delete txDataNoFrom['from']
    st.plan(1)

    var tx = new FakeTransaction(txDataNoFrom)
    st.equal(utils.bufferToHex(tx.from), txData.from)
  })
})


test/index.js
const argv = require('minimist')(process.argv.slice(2))

if (argv.a) {
  require('./api.js')
} else if (argv.t) {
  require('./transactionRunner.js')
} else {
  require('./api.js')
  require('./transactionRunner.js')
}


fake.js
'use strict'

const Transaction = require('./index.js')
const moacUtil = require('moacjs-util')

/**
 * Creates a new transaction object that doesn't need to be signed
 * @constructor
 * @class {Buffer|Array} data a transaction can be initialized with either a buffer containing the RLP serialized transaction or an array of buffers relating to each of the tx Properties, listed in order below in the exmple. Or lastly an Object containing the Properties of the transaction like in the Usage example
 *
 * For Object and Arrays each of the elements can either be a Buffer, a hex-prefixed (0x) String , Number, or an object with a toBuffer method such as Bignum
 * @example
 * var rawTx = {
 *   nonce: '0x00',
 *   gasPrice: '0x09184e72a000',
 *   gasLimit: '0x2710',
 *   to: '0x0000000000000000000000000000000000000000',
 *   value: '0x00',
 *   data: '0x7f7465737432000000000000000000000000000000000000000000000000000000600057',
 *   v: '0x1c',
 *   r: '0x5e1d3a76fbf824220eafc8c79ad578ad2b67d01b0c2425eb1f1347e8f50882ab',
 *   s '0x5bd428537f05f9830e93792f90ea6a3e2d1ee84952dd96edbae9f658f831ab13'
 * };
 * var tx = new Transaction(rawTx);
 * @prop {Buffer} raw The raw rlp decoded transaction
 * @prop {Buffer} nonce
 * @prop {Buffer} to the to address
 * @prop {Buffer} value the amount of ether sent
 * @prop {Buffer} data this will contain the data of the message or the init of a contract
 * @prop {Buffer} v EC recovery ID
 * @prop {Buffer} r EC signature parameter
 * @prop {Buffer} s EC signature parameter
 */
module.exports = class FakeTransaction extends Transaction {
  constructor (data) {
    super(data)

    var self = this

    /**
     * @prop {Buffer} from (read/write) Set from address to bypass transaction signing.
     */
    Object.defineProperty(this, 'from', {
      enumerable: true,
      configurable: true,
      get: this.getSenderAddress.bind(self),
      set: function (val) {
        if (val) {
          self._from = moacUtil.toBuffer(val)
        }
      }
    })

    this.from = data.from
  }

  /**
   * Computes a sha3-256 hash of the serialized tx, using the sender address to generate a fake signature.
   * @param {Boolean} [includeSignature=true] whether or not to inculde the signature
   * @return {Buffer}
   */
  hash (includeSignature = true) {
    if (includeSignature && this._from && this._from.toString('hex') !== '') {
      // include a fake signature using the from address as a private key
      let fakeKey = Buffer.concat([this._from, this._from.slice(0, 12)])
      this.sign(fakeKey)
    }

    return super.hash(includeSignature)
  }
}



utils.js
const BN = require('bn.js')
const numberToBN = require('number-to-bn')
const _ = require('underscore')
const Buffer = require('safe-buffer').Buffer
const moacUtil = require('moacjs-util')

/**
 * This function is to resolve the issue
 * https://github.com/ethereum/web3.js/issues/1170
 */
function makeEven (hex) {
  if (hex.length % 2 === 1) {
    hex = hex.replace('0x', '0x0')
  }
  return hex
}

/**
 * Takes an input and transforms it into an BN
 *
 * @method toBN
 * @param {Number|String|BN} number, string, HEX string or BN
 * @return {BN} BN
 */
function toBN (number) {
  try {
    return numberToBN.apply(null, arguments)
  } catch (e) {
    throw new Error(e + ' Given value: "' + number + '"')
  }
}

/**
 * Converts value to it's hex representation
 *
 * @method numberToHex
 * @param {String|Number|BN} value
 * @return {String}
 */
function numberToHex (value) {
  if (_.isNull(value) || _.isUndefined(value)) {
    return value
  }

  if (!isFinite(value) && !moacUtil.isHexString(value)) {
    throw new Error('Given input "' + value + '" is not a number.')
  }

  var number = toBN(value)
  var result = number.toString(16)

  return number.lt(new BN(0)) ? '-0x' + result.substr(1) : '0x' + result
}

// To fix an error of 2 leading 0s
function trimLeadingZero (hex) {
  while (hex && hex.startsWith('0x00')) {
    hex = '0x' + hex.slice(4)
  }
  return hex
}

/**
 * Attempts to turn a value into a `Buffer`. As input it supports `Buffer`, `String`, `Number`,
 * null/undefined, `BN` and other objects with a `toArray()` method.
 * @param {*} v the value
 */
function toBuffer (v) {
  if (!Buffer.isBuffer(v)) {
    if (Array.isArray(v)) {
      v = Buffer.from(v)
    } else if (typeof v === 'string') {
      if (moacUtil.isHexString(v)) {
        v = Buffer.from(moacUtil.padToEven(moacUtil.stripHexPrefix(v)), 'hex')
      } else {
        v = Buffer.from(v)
      }
    } else if (typeof v === 'number') {
      v = moacUtil.intToBuffer(v)
    } else if (v === null || v === undefined) {
      v = Buffer.allocUnsafe(0)
    } else if (v.toArray) {
      // converts a BN to a Buffer
      v = Buffer.from(v.toArray())
    } else {
      throw new Error('invalid type')
    }
  }
  return v
}

/**
 * Converts a `Buffer` into a hex `String`
 * @param {Buffer} buf
 * @return {String}
 */
function bufferToHex (buf) {
  buf = toBuffer(buf)
  return '0x' + buf.toString('hex')
}

module.exports = {
  toBuffer: toBuffer,
  bufferToHex: bufferToHex,
  trimLeadingZero: trimLeadingZero,
  makeEven: makeEven,
  toBN: toBN,
  numberToHex: numberToHex
}


index.js
'use strict'
const moacUtil = require('moacjs-util')
const fees = require('ethereum-common/params.json')
const BN = moacUtil.BN
const RLP = require('eth-lib/lib/rlp')
const Bytes = require('eth-lib/lib/bytes')
const Hash = require('eth-lib/lib/hash')
const secp256k1 = require('secp256k1')
const numberToBN = require('number-to-bn')
const _ = require('underscore')
const Buffer = require('safe-buffer').Buffer

// secp256k1n/2
const N_DIV_2 = new BN('7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46681b20a0', 16)

/**
 * ECDSA sign
 * @param {Buffer} msgHash
 * @param {Buffer} privateKey
 * @return {Object}
 */
function ecsign (msgHash, privateKey) {
  // Convert the input string to Buffer
  if (typeof msgHash === 'string') {
    if (moacUtil.isHexString(msgHash)) {
      msgHash = Buffer.from(makeEven(moacUtil.stripHexPrefix(msgHash)), 'hex')
    }
  }

  var privateKeyBuf = new Buffer(privateKey, 'hex')

  var sig = secp256k1.sign(msgHash, privateKeyBuf)

  var ret = {}
  ret.r = sig.signature.slice(0, 32)
  ret.s = sig.signature.slice(32, 64)
  ret.v = sig.recovery + 27

  return ret
}

/**
 * Creates a new transaction object.
 *
 * @example
 * var rawTx = {
  nonce: '0x00',
  gasPrice: '0x09184e72a000',
  gasLimit: '0x2710',
  to: '0x0000000000000000000000000000000000000000',
  value: '0x00',
  data: '0x7f7465737432000000000000000000000000000000000000000000000000000000600057',
  shardingFlag: 0,
  // chainId - mainnet: 99, testnet: 101
  chainId: 101
 * };
 * var tx = new Transaction(rawTx);
 *
 * @class
 * @param {Buffer | Array | Object} data a transaction can be initialized with either a buffer containing the RLP serialized transaction or an array of buffers relating to each of the tx Properties, listed in order below in the exmple.
 *
 * Or lastly an Object containing the Properties of the transaction like in the Usage example.
 *
 * For Object and Arrays each of the elements can either be a Buffer, a hex-prefixed (0x) String , Number, or an object with a toBuffer method such as Bignum
 *
 * @property {Buffer} raw The raw rlp encoded transaction
 * @param {Buffer} data.nonce nonce number
 * @param {Buffer} data.gasLimit transaction gas limit
 * @param {Buffer} data.gasPrice transaction gas price
 * @param {Buffer} data.to to the to address
 * @param {Buffer} data.value the amount of moac sent
 * @param {Buffer} data.data this will contain the data of the message or the init of a contract
 * @param {Buffer} data.shardingFlag choose sharding or not
 * @param {Number} data.chainId moac chainId - mainnet: 99, testnet: 101
 * */
class Transaction {
  constructor (data) {
    data = data || {}
    // Define Properties
    const fields = [{
      name: 'nonce',
      length: 32,
      allowLess: true,
      default: new Buffer([])
    }, {
      name: 'gasPrice',
      length: 32,
      allowLess: true,
      default: new Buffer([])
    }, {
      name: 'gasLimit',
      alias: 'gas',
      length: 32,
      allowLess: true,
      default: new Buffer([])
    }, {
      name: 'to',
      allowZero: true,
      length: 20,
      default: new Buffer([])
    }, {
      name: 'value',
      length: 32,
      allowLess: true,
      default: new Buffer([])
    }, {
      name: 'data',
      alias: 'input',
      allowZero: true,
      default: new Buffer([])
    }, {
      name: 'v',
      allowZero: true,
      default: new Buffer([0x1c])
    }, {
      name: 'r',
      length: 32,
      allowZero: true,
      allowLess: true,
      default: new Buffer([])
    }, {
      name: 's',
      length: 32,
      allowZero: true,
      allowLess: true,
      default: new Buffer([])
    }]

    /**
     * Returns the rlp encoding of the transaction
     * @method serialize
     * @return {Buffer}
     * @memberof Transaction
     * @name serialize
     * @see {@link https://github.com/wanpixiaozi/moacjs-tx/blob/master/docs/index.md#defineproperties|moacjs-util}
     */
    /**
     * Returns the transaction in JSON format
     * @method toJSON
     * @return {Array | String}
     * @memberof Transaction
     * @name toJSON
     * @see {@link https://github.com/wanpixiaozi/moacjs-tx/blob/master/docs/index.md#defineproperties|moacjs-util}
     */
    // attached serialize
    moacUtil.defineProperties(this, fields, data)

    /**
     * @property {Buffer} from (read only) sender address of this transaction, mathematically derived from other parameters.
     * @name from
     * @memberof Transaction
     */
    Object.defineProperty(this, 'from', {
      enumerable: true,
      configurable: true,
      get: this.getSenderAddress.bind(this)
    })

    // calculate chainId from signature
    let sigV = moacUtil.bufferToInt(this.v)
    let chainId = Math.floor((sigV - 35) / 2)
    if (chainId < 0) chainId = 0

    // set chainId
    this._chainId = chainId || data.chainId || 0
    this._homestead = true

    this.tx = data // used for moac sign process
  }

  /**
   * If the tx's `to` is to the creation address
   * @return {Boolean}
   */
  toCreationAddress () {
    return this.to.toString('hex') === ''
  }

  /**
   * Computes a sha3-256 hash of the serialized tx
   * @param {Boolean} [includeSignature=true] whether or not to include the signature
   * @return {Buffer}
   */
  hash (includeSignature) {
    if (includeSignature === undefined) includeSignature = true

    // EIP155 spec:
    // when computing the hash of a transaction for purposes of signing or recovering,
    // instead of hashing only the first six elements (ie. nonce, gasprice, startgas, to, value, data),
    // hash nine elements, with v replaced by CHAIN_ID, r = 0 and s = 0

    let items
    if (includeSignature) {
      items = this.raw
    } else {
      if (this._chainId > 0) {
        const raw = this.raw.slice()
        this.v = this._chainId
        this.r = 0
        this.s = 0
        items = this.raw
        this.raw = raw
      } else {
        items = this.raw.slice(0, 6)
      }
    }

    // create hash
    return moacUtil.rlphash(items)
  }

  /**
   * returns chain ID
   * @return {Buffer}
   */
  getChainId () {
    return this._chainId
  }

  /**
   * returns the sender's address
   * @return {Buffer}
   */
  getSenderAddress () {
    if (this._from) {
      return this._from
    }
    const pubkey = this.getSenderPublicKey()
    this._from = moacUtil.publicToAddress(pubkey)
    return this._from
  }

  /**
   * returns the public key of the sender
   * @return {Buffer}
   */
  getSenderPublicKey () {
    if (!this._senderPubKey || !this._senderPubKey.length) {
      if (!this.verifySignature()) throw new Error('Invalid Signature')
    }
    return this._senderPubKey
  }

  /**
   * Determines if the signature is valid
   * @return {Boolean}
   */
  verifySignature () {
    const msgHash = this.hash(false)
    // All transaction signatures whose s-value is greater than secp256k1n/2 are considered invalid.
    if (this._homestead && new BN(this.s).cmp(N_DIV_2) === 1) {
      return false
    }

    try {
      let v = moacUtil.bufferToInt(this.v)
      if (this._chainId > 0) {
        v -= this._chainId * 2 + 8
      }
      this._senderPubKey = moacUtil.ecrecover(msgHash, v, this.r, this.s)
    } catch (e) {
      return false
    }

    return !!this._senderPubKey
  }

  /**
   * sign a transaction with a given private key
   * @param {Buffer} privateKey Must be 32 bytes in length
   * eth old func, moac never use it
   *
   * sign (privateKey) {
   *  const msgHash = this.hash(false)
   *  const sig = moacUtil.ecsign(msgHash, privateKey)
   *  if (this._chainId > 0) {
   *     sig.v += this._chainId * 2 + 8
   *  }
   *  Object.assign(this, sig)
   * }
   **/

  /*
   * A simple signTransaction function to sign
   * the input TX with private key.
   * Input:
   * tx - a JSON format object contains the input TX info
   * privateKey - a string format of the private key
   * Output:
   * rawTransaction - a String, can be used with
   *                  chain3.mc.sendRawTransaction
   *
   *
  */
  sign (privateKey) {
    const tx = this.tx
    // Check the input fields of the tx
    if (tx.chainId < 1) {
      return new Error('"Chain ID" is invalid')
    }

    if (!tx.gas && !tx.gasLimit) {
      return new Error('"gas" is missing')
    }

    if (tx.nonce < 0 ||
      tx.gasLimit < 0 ||
      tx.gasPrice < 0 ||
      tx.chainId < 0) {
      return new Error('Gas, gasPrice, nonce or chainId is lower than 0')
    }
    // Sharding Flag only accept the
    // If input has not sharding flag, set it to 0 as global TX.
    if (tx.shardingFlag === undefined) {
      this.shardingFlag = 0
    }
    try {
      // Make sure all the number fields are in HEX format

      var transaction = tx
      transaction.to = tx.to || '0x' // Can be zero, for contract creation
      transaction.data = tx.data || '0x' // can be zero for general TXs
      transaction.value = tx.value || '0x' // can be zero for contract call
      transaction.chainId = numberToHex(tx.chainId)
      transaction.shardingFlag = numberToHex(tx.shardingFlag)
      transaction.systemContract = '0x0' // System contract flag, always = 0
      transaction.via = tx.via || '0x' // vnode subchain address

      const rlpEncoded = RLP.encode([
        Bytes.fromNat(transaction.nonce),
        Bytes.fromNat(transaction.systemContract),
        Bytes.fromNat(transaction.gasPrice),
        Bytes.fromNat(transaction.gasLimit),
        transaction.to.toLowerCase(),
        Bytes.fromNat(transaction.value),
        transaction.data,
        Bytes.fromNat(transaction.shardingFlag),
        transaction.via.toLowerCase(),
        Bytes.fromNat(transaction.chainId),
        '0x',
        '0x'])

      const hash = Hash.keccak256(rlpEncoded)

      // for MOAC, keep 9 fields instead of 6
      const vPos = 9
      // Sign the hash with the private key to produce the
      // V, R, S
      const newsign = ecsign(hash, moacUtil.stripHexPrefix(privateKey))
      const rawTx = RLP.decode(rlpEncoded).slice(0, vPos + 3)

      // Replace the V field with chainID info
      const newV = newsign.v + 8 + transaction.chainId * 2

      // Add trimLeadingZero to avoid '0x00' after makeEven
      // don't allow uneven r,s,v values
      rawTx[vPos] = trimLeadingZero(makeEven(bufferToHex(newV)))
      rawTx[vPos + 1] = trimLeadingZero(makeEven(bufferToHex(newsign.r)))
      rawTx[vPos + 2] = trimLeadingZero(makeEven(bufferToHex(newsign.s)))
      return RLP.encode(rawTx)
    } catch (e) {
      return e
    }
  }

  /**
   * The amount of gas paid for the data in this tx
   * @return {BN}
   */
  getDataFee () {
    const data = this.raw[5]
    const cost = new BN(0)
    for (let i = 0; i < data.length; i++) {
      data[i] === 0 ? cost.iaddn(fees.txDataZeroGas.v) : cost.iaddn(fees.txDataNonZeroGas.v)
    }
    return cost
  }

  /**
   * the minimum amount of gas the tx must have (DataFee + TxFee + Creation Fee)
   * @return {BN}
   */
  getBaseFee () {
    const fee = this.getDataFee().iaddn(fees.txGas.v)
    if (this._homestead && this.toCreationAddress()) {
      fee.iaddn(fees.txCreation.v)
    }
    return fee
  }

  /**
   * the up front amount that an account must have for this transaction to be valid
   * @return {BN}
   */
  getUpfrontCost () {
    return new BN(this.gasLimit)
      .imul(new BN(this.gasPrice))
      .iadd(new BN(this.value))
  }

  /**
   * validates the signature and checks to see if it has enough gas
   * @param {Boolean} [stringError=false] whether to return a string with a description of why the validation failed or return a Boolean
   * @return {Boolean|String}
   */
  validate (stringError) {
    const errors = []
    if (!this.verifySignature()) {
      errors.push('Invalid Signature')
    }

    if (this.getBaseFee().cmp(new BN(this.gasLimit)) > 0) {
      errors.push([`gas limit is too low. Need at least ${this.getBaseFee()}`])
    }

    if (stringError === undefined || stringError === false) {
      return errors.length === 0
    } else {
      return errors.join(' ')
    }
  }
}
/**
 */
function makeEven (hex) {
  if (hex.length % 2 === 1) {
    hex = hex.replace('0x', '0x0')
  }
  return hex
}

/**
 * Takes an input and transforms it into an BN
 *
 * @method toBN
 * @param {Number|String|BN} number, string, HEX string or BN
 * @return {BN} BN
 */
function toBN (number) {
  try {
    return numberToBN.apply(null, arguments)
  } catch (e) {
    throw new Error(e + ' Given value: "' + number + '"')
  }
}

/**
 * Converts value to it's hex representation
 *
 * @method numberToHex
 * @param {String|Number|BN} value
 * @return {String}
 */
function numberToHex (value) {
  if (_.isNull(value) || _.isUndefined(value)) {
    return value
  }

  if (!isFinite(value) && !moacUtil.isHexString(value)) {
    throw new Error('Given input "' + value + '" is not a number.')
  }

  var number = toBN(value)
  var result = number.toString(16)

  return number.lt(new BN(0)) ? '-0x' + result.substr(1) : '0x' + result
}

// To fix an error of 2 leading 0s
function trimLeadingZero (hex) {
  while (hex && hex.startsWith('0x00')) {
    hex = '0x' + hex.slice(4)
  }
  return hex
}

/**
 * Attempts to turn a value into a `Buffer`. As input it supports `Buffer`, `String`, `Number`,
 * null/undefined, `BN` and other objects with a `toArray()` method.
 * @param {*} v the value
 */
function toBuffer (v) {
  if (!Buffer.isBuffer(v)) {
    if (Array.isArray(v)) {
      v = Buffer.from(v)
    } else if (typeof v === 'string') {
      if (moacUtil.isHexString(v)) {
        v = Buffer.from(moacUtil.padToEven(moacUtil.stripHexPrefix(v)), 'hex')
      } else {
        v = Buffer.from(v)
      }
    } else if (typeof v === 'number') {
      v = moacUtil.intToBuffer(v)
    } else if (v === null || v === undefined) {
      v = Buffer.allocUnsafe(0)
    } else if (v.toArray) {
      // converts a BN to a Buffer
      v = Buffer.from(v.toArray())
    } else {
      throw new Error('invalid type')
    }
  }
  return v
}

/**
 * Converts a `Buffer` into a hex `String`
 * @param {Buffer} buf
 * @return {String}
 */
function bufferToHex (buf) {
  buf = toBuffer(buf)
  return '0x' + buf.toString('hex')
}

module.exports = Transaction
